Sweetener.Logging/Loggers/Logger.cs

```diff
@@ -0,0 +1,162 @@
+
+            FormatProvider = formatProvider;
+            MinimumLevel   = minimumLevel;
+            _logPattern    = new LogPattern(pattern);
+        }
+
+        /// <summary>
+        /// Releases all resources used by the <see cref="Logger"/> object.
+        /// </summary>
+        /// <remarks>
+        /// Call <see cref="Dispose"/> when you are finished using the <see cref="Logger"/>.
+        /// The <see cref="Dispose"/> method leaves the <see cref="Logger"/> in an unusable state.
+        /// After calling <see cref="Dispose"/>, you must release all references to the
+        /// <see cref="Logger"/> so the garbage collector can reclaim the memory that the
+        /// <see cref="Logger"/> was occupying.
+        /// </remarks>
+        public void Dispose()
+        {
+            Dispose(true);
+
+            // Still suppress, in case any derived classes use a finalizer
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Releases the unmanaged resources used by the <see cref="Logger"/> and
+        /// optionally releases the managed resources.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// This method is called by <see cref="Dispose"/> and possible Finalize in a
+        /// derived class. By default, this method specifies the <paramref name="disposing"/>
+        /// parameter as <c>true</c>. Any Finalize implementation specifies the
+        /// <paramref name="disposing"/> parameter as <c>false</c>.
+        /// </para>
+        /// <para>
+        /// When the <paramref name="disposing"/> parameter is <c>true</c>, this method
+        /// releases all resources held by any managed objects that this <see cref="Logger"/>
+        /// references. This method invokes the <see cref="Dispose"/> method of each referenced object.
+        /// </para>
+        /// </remarks>
+        /// <param name="disposing">
+        /// <c>true</c> to release both managed and unmanaged resources; <c>false</c>
+        /// to release only unmanaged resources.
+        /// </param>
+        protected virtual void Dispose(bool disposing)
+        {
+            // In the base class, this flag is used to short circuit calls before Log(...)
+            // We don't want method calls on disabled loggers to not throw exceptions
+            // because the log request was below the minimum log level!
+            if (!_disposed)
+                _disposed = true;
+        }
+
+        /// <summary>
+        /// Writes the message to the log.
+        /// </summary>
+        /// <param name="message">The value to be logged.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="message"/> is <c>null</c>.</exception>
+        protected abstract void WriteLine(string message);
+
+        private void Log(LogLevel level, string message)
+            => WriteLine(_logPattern.Format(FormatProvider, level, message));
+
+        private void ThrowIfDisposed()
```

Comment:
This is a pattern used in [StreamWriter.cs](https://github.com/dotnet/corefx/blob/6f9570e012054bf407778c0b07632e7733ecbb13/src/Common/src/CoreLib/System/IO/StreamWriter.cs#L1092) from CoreFx